crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,1039 @@
+    fn get_type_root(&self) -> &DataTypeRoot {
+        &self.type_root
+    }
+
+    /// Returns whether the root of the type equals to the type_root or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L75>
+    fn is(&self, type_root: &DataTypeRoot) -> bool {
+        &self.type_root == type_root
+    }
+
+    /// Returns whether the family type of the type equals to the family or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L103>
+    fn is_with_family(&self, family: DataTypeFamily) -> bool {
+        self.type_root.families().contains(family)
+    }
+
+    /// Returns whether the root of the type equals to at least on of the type_roots or not.
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L84>
+    fn is_any_of(&self, type_roots: &[DataTypeRoot]) -> bool {
+        type_roots.iter().any(|tr: &DataTypeRoot| self.is(tr))
+    }
+
+    /// Returns whether the root of the type is part of at least one family of the families or not.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L94>
+    fn is_any_of_family(&self, families: &[DataTypeFamily]) -> bool {
+        families
+            .iter()
+            .any(|f: &DataTypeFamily| self.is_with_family(f.clone()))
+    }
+
+    /// Returns a deep copy of this type with possibly different nullability.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L113>
+    fn copy(&self, is_nullable: bool) -> Self {
+        Self {
+            is_nullable,
+            type_root: self.type_root.clone(),
+        }
+    }
+
+    /// Returns a deep copy of this type. It requires an implementation of {@link #copy(boolean)}.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L120>
+    fn copy_with_nullable(&self) -> Self {
+        self.copy(self.is_nullable)
+    }
+
+    /// Compare two data types without nullable.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L129>
+    fn copy_ignore_nullable(&self) -> Self {
+        self.copy(false)
+    }
+
+    fn serialize_json(&self) -> String {
```

Comment:
Ideally, we don't need to provide API like this directly. If we have special requirement for `Serialize`, we need to implement by hand instead of using `#[derive(Serialize)]`